### ISSUE TITLE
nested if incorrectly raises inequalities error.

### ISSUE THREAD

**Issue by sgparry** [maths/moodle-qtype_stack#324]
e.g.
f(x) := if x < 0 then (if x < 1 then 1 else 2) else 3;
gives
You appear to have "chained inequalities" e.g. a<b<c. You need to connect individual inequalities with logical operations such as and or or.

**Comment by aharjula** [comment]
Oh my... Chris, we really need to use a proper parser to distinguish true inequalities from logic. Luckily, that can be written in pieces to skip that false positive from that test. Counting chars without contextual awareness just does not work you need a parse tree presentation where you can check that no inequality node has inequalities as its direct children. 

```
g(x):=if x < 1 then 1 else 2;
f(x):=if x < 0 then g(x) else 3;
```

Chris, I see no way of fixing [this](https://github.com/maths/moodle-qtype_stack/blob/e471141c8adbd9150860a93040dd625708fefe5f/stack/cas/casstring.class.php#L1252.L1273) without using a tree presentation of the expression. So we will need to start building such.

As that test has been in the system for quite some time I don't believe it is an acute problem that would break existing materials but I for one would prefer to be able to write logical expressions without having to work around false positives coming from syntax checkers.

**Comment by sangwinc** [comment]
I agree Matti,

What, for some time, has been a simple check designed for student's expressions will not work for more complex logic and we need a full parser for this.  

To stop this breaking question authors in the mean time we could only apply this check to student's expressions.  This means teachers won't get the "error" message, but they will also not be invalidated in the situation described above.  I don't intend students to enter this kind of logic.

**Comment by sgparry** [comment]
That would suit me as a short term workaround..

**Comment by sgparry** [comment]
HELP! is there somewhere in the php I can comment this check out please? I cannot even get expressions with logical connectives in to parse properly! This is really ruining my day, again.

**Comment by sangwinc** [comment]
Yes, 
The only place this check is used is here: https://github.com/maths/moodle-qtype_stack/blob/e471141c8adbd9150860a93040dd625708fefe5f/stack/cas/casstring.class.php#L892

**Comment by sgparry** [comment]
I kind of cheated and got the function to return true, but same effect. Thanks anyway, I'm making progress now.

**Comment by aharjula** [comment]
Chris, do you think you could work with this kind of tool:

```
$tree = mysticalParser("x < y + 1 <= z"); 
// => Inequality('<',
//                Var(x),
//                Inequality('<=',
//                            Sum(Var(y),Num(1)),
//                            Var(z)))

function chainedTest($node) {
 if ($node instanceof Inequality) {
  if ($node.lhs instanceof Inequality || $node.rhs instanceof Inequality) {
   return true;
  }
 }
 return false;
}

$badNodes = $tree.findNodes(chainedTest);
if (count($badOnes) > 0)
 argh! there is an chained inequality between $badNodes[0].startIndex and $badNodes[0].endIndex...
```

If so then I could maybe provide that by trying to compile the parser I showed you in June to PHP and build some logic for filtering on top of it. Maybe that could be usable early next year.

Problem is that that parsing will be expensive for just one validation and we should try to use it fully i.e. move as many of the current tests we have to use that generated tree instead of evaluating regexps (although they are often cheap they are not free and are much more complex to read on the code side).

**Comment by sangwinc** [comment]
I agree we need to develop a full parser at some point in the future.

I've committed some changes for review which I hope will allow teachers to write questions with logic in.  As we expand the range of available features we have to accommodate these, and some of the very simplistic checks we used in the past will be no longer sufficient.  These changes mean teachers won't get errors about chained inequalities which they used to.  From the student perspective this should be unchanged.
